### PR TITLE
Issue385 timeout

### LINKED ISSUE #385: Timeout handling by Simulator for Optimica and JModelica not working
In `development/optimica_run.template` the process timeout is managed this way:
```
    p = multiprocessing.Process(target=target, args=(0, return_dict))
    start = time.time()
    p.start()
    if timeout > 0:
        p.join(timeout)
    else:
        p.join()
    cpu_time = time.time()-start
    if p.exitcode != 0:
        if cpu_time >= timeout:
            msg = "Process timeout"
        else:
            msg = "Process terminated by signal {}".format(-p.exitcode)
        raise RuntimeError(msg)
```
- To disable the timeout, a negative value is used (-1). However, if the child process execution fails, the assertion `cpu_time >= timeout` will always return true, hindering the real process exit code.
- The timeout parameter of the `join()` method is the timeout of the method itself, not of the process being joined. So when a timeout is set and hit, the child process is not terminated and `_simulate` method does not return. The simulation continues despite a runtime error for process timeout being logged.

EDIT: After correcting the previous issue, another one appears from `_BaseSimulator._runSimulation`. Setting a positive timeout value and simulating a model with Optimica yields the following error stream in the terminal.
```
*** Error: Killing simulation in ..
*** Error: Killing simulation in ..
```
However the simulation is carried out until the end in the background (`"success": true,` in `*_buildingspy.json`). It is suspected that the `kill()` method used in `_BaseSimulator._runSimulation` does not terminate the nested subprocess started with `multiprocessing.Process()` (within `development/optimica_run.template`). And the `kill()` method is always used (instead of `terminate`) if `_showProgressBar = False` (the default) due to the following.

```
                    if elapsedTime > timeout:
                        # First, terminate the process. Then, if it is still
                        # running, kill the process

                        if self._showProgressBar and not killedProcess:
                            killedProcess = True
                            # This output needed because of the progress bar
                            sys.stdout.write("\n")
                            self._reporter.writeError("Terminating simulation in " +
                                                      directory + ".")
                            pro.terminate()
                        else:
                            self._reporter.writeError("Killing simulation in " +
                                                      directory + ".")
                            pro.kill()
```
We could either

- recursively kill the child processes using `process = psutil.Process(proc.pid); for p in process.children(recursive=True): p.kill()`, or
- (preferred solution) only call `kill()` for dymosim and pass through the timeout as an argument to `template.render()` for Optimica and JModelica so that the termination originates from the child process (with proper error log) and the main process can gracefully exit when its children are terminated.

### PR DESCRIPTION
**Pull request by AntoineGautier**
This closes #385

### CODE REVIEW

**Comment by AntoineGautier** [comment]
Before merging we should further validate the following code snippet.
```
                    if elapsedTime > timeout:
                        # First, terminate the process. Then, if it is still
                        # running, kill the process

                        if self._showProgressBar and not killedProcess:
                            killedProcess = True
                            # This output needed because of the progress bar
                            sys.stdout.write("\n")
                            self._reporter.writeError("Terminating simulation in " +
                                                      directory + ".")
                            pro.terminate()
                        else:
                            self._reporter.writeError("Killing simulation in " +
                                                      directory + ".")
                            pro.kill()
```
The code does not seem to do what the comment states. Indeed, if `_showProgressBar=False` then the process is always killed, never terminated. I already modified this part by putting `killedProcess = True` outside of `if self._showProgressBar and not killedProcess` since the two branches of the assertion should update `killedProcess` I believe.

**Comment by AntoineGautier** [comment]
@mwetter This is ready for a review. Please consider also my comment here https://github.com/lbl-srg/BuildingsPy/pull/386#issuecomment-722127533.

**Review by mwetter — APPROVED** [review]
@AntoineGautier : Please merge the PR if you agree with the changes and all tests pass. See commit messages.

**Comment by AntoineGautier** [comment]
TODO

- [ ] Add a test for the timeout mechanism for all simulation tools.

File: buildingspy/simulate/base_simulator.py [hunk 1]
```diff
@@ -407,17 +408,19 @@ def _runSimulation(self, cmd, timeout, directory, env=None):
...
-                                sys.stdout.write("\n")
+                            if not terminatedProcess:
+                                if self._showProgressBar:
+                                    # This output needed because of the progress bar
+                                    sys.stdout.write("\n")
                                 self._reporter.writeError("Terminating simulation in " +
                                                           directory + ".")
                                 pro.terminate()
+                                terminatedProcess = True
                             else:
```
[the remark this answers is not in the document]
Comment (AntoineGautier) [reply buildingspy/simulate/base_simulator.py]: @mwetter My mistake (this was a Monday morning comment). The `else` branch gets indeed evaluated if `terminatedProcess == True` and `pro.poll()` is still `None`. However, testing it on a minimal example shows that `pro.poll()` is still `None` after `pro.terminate()` as the process needs some time to terminate. I proposed a `terminate_timeout` to handle that in https://github.com/lbl-srg/BuildingsPy/pull/386/commits/e5dcdc0f8cc9cddcc384ddd59c5a9ede1c2158fd

[the remark this answers is not in the document]
Comment (mwetter) [reply buildingspy/simulate/base_simulator.py]: @AntoineGautier : No longer supporting python 2 is fine.
I implemented a Modelica function that sleeps in 193153a through a C function call, as this seems easier than calling `Modelica.Utilities.System.getTime` and computing the differential time.

**Comment by AntoineGautier** [comment]
@mwetter This is ready to merge (thanks for the tailored sleep function: awesome!).
I did not refactor to use the `timeout` argument from recent versions of `subprocess`. So the package is still compatible with Python 2. However, I detected what seemed to me an error in the specification of the Python version, see https://github.com/lbl-srg/BuildingsPy/pull/386/commits/5d1ff53ad94425db7c9a5d0bcd4fe4227ec8ea6c#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R34.

**Comment by mwetter** [comment]
@AntoineGautier : Thanks for the version fix and the unit test. This looks good and I will merge it.

**Pull request merged**